projects/plutus/src/ltc_afe.c

```diff
@@ -0,0 +1,310 @@
+  configuration_cmd[configuration_index + 0] = (uint8_t)(wrcfg >> 8);
+  configuration_cmd[configuration_index + 1] = (uint8_t)(wrcfg & 0xFF);
+
+  uint16_t cmd_pec = crc15_calculate(configuration_cmd, 2);
+  configuration_cmd[configuration_index + 2] = (uint8_t)(cmd_pec >> 8);
+  configuration_cmd[configuration_index + 3] = (uint8_t)(cmd_pec & 0xFF);
+
+  configuration_index += 4;
+  // send CFGR registers starting with the bottom slave in the stack
+  for (uint8_t device = PLUTUS_AFE_DEVICES_IN_CHAIN; device > 0; --device) {
+    uint8_t current_device = PLUTUS_AFE_DEVICES_IN_CHAIN - device;
+    uint8_t enable = gpio_enable_pins;
+    uint16_t undervoltage = 0;
+    uint16_t overvoltage = 0;
+    uint16_t cells_to_discharge = 0;
+
+    for (uint8_t cell = 0; cell < LTC6804_CELLS_PER_DEVICE; ++cell) {
+      uint16_t cell_index = current_device * LTC6804_CELLS_PER_DEVICE + cell;
+      cells_to_discharge |= (s_discharging_cells[cell_index] << cell);
+    }
+    LTCAFEDischargeTimeout timeout = LTC_AFE_DISCHARGE_TIMEOUT_1_MIN;
+
+    // CFGR0
+    configuration_cmd[configuration_index + 0] = enable;
+    // (adc mode enum + 1) > 3:
+    //    - true: CFGR0[0] = 1
+    //    - false: CFGR0[0] = 0
+    // CFGR0: bit0 is the ADC Mode
+    configuration_cmd[configuration_index + 0] |= ((afe->adc_mode + 1) > 3);
+    configuration_cmd[configuration_index + 0] |= LTC6804_SWTRD;
+
+    // CFGR1: VUV[7...0]
+    configuration_cmd[configuration_index + 1] = (undervoltage & 0xFF);
+
+    // CFGR2: VUV[11...8] in bit3, ..., bit0
+    configuration_cmd[configuration_index + 2] = ((undervoltage >> 8) & 0x0F);
+    // CFGR2: VOV[3...0] in bit7, ..., bit4
+    configuration_cmd[configuration_index + 2] |= ((overvoltage << 4) & 0xF0);
+
+    // CFGR3: VOV[11...4]
+    configuration_cmd[configuration_index + 3] = ((overvoltage >> 4) & 0xFF);
+
+    // CFGR4: DCC8, ..., DCC1
+    configuration_cmd[configuration_index + 4] = (cells_to_discharge & 0xFF);
+
+    // CFGR5: DCC12, ..., DCC9
+    configuration_cmd[configuration_index + 5] = ((cells_to_discharge >> 8) & 0x0F);
+    // CFGR5: DCTO5, ... DCTO0 in bit7, ..., bit4
+    configuration_cmd[configuration_index + 5] |= (timeout << 4);
+
+    // adjust the offset to point to the start of slave's configuration
+    uint8_t *offset = configuration_cmd + (2 + 2) + ((6 + 2) * current_device);
+    uint16_t cfgr_pec = crc15_calculate(offset, 6);
+    configuration_cmd[configuration_index + 6] = (uint8_t)(cfgr_pec >> 8);
+    configuration_cmd[configuration_index + 7] = (uint8_t)(cfgr_pec & 0xFF);
+
+    configuration_index += 8;
+  }
+
+  prv_wakeup_idle(afe);
+
+  return spi_exchange(afe->spi_port, configuration_cmd, cfg_len, NULL, 0);
+}
+
+static StatusCode prv_read_config(const LTCAFESettings *afe, uint8_t *configuration_registers) {
+  StatusCode status = STATUS_CODE_OK;
+
+  prv_wakeup_idle(afe);
+
+  uint8_t received_data[(6 + 2) * PLUTUS_AFE_DEVICES_IN_CHAIN] = { 0 };
```

Comment:
Would it be possible to use structs or macros to give more context to these size calculations? It's pretty confusing, especially when it comes to indexing.